framework/src/main/java/org/checkerframework/common/returnsreceiver/ReturnsReceiverAnnotatedTypeFactory.java

```diff
@@ -0,0 +1,98 @@
+
+    /**
+     * Create a new {@code ReturnsReceiverAnnotatedTypeFactory}.
+     *
+     * @param checker the type-checker associated with this factory
+     */
+    public ReturnsReceiverAnnotatedTypeFactory(BaseTypeChecker checker) {
+        super(checker);
+        THIS_ANNOTATION = AnnotationBuilder.fromClass(elements, This.class);
+        this.postInit();
+    }
+
+    @Override
+    protected Set<Class<? extends Annotation>> createSupportedTypeQualifiers() {
+        return getBundledTypeQualifiers(BottomThis.class, UnknownThis.class, This.class);
+    }
+
+    @Override
+    protected TypeAnnotator createTypeAnnotator() {
+        return new ListTypeAnnotator(
+                new ReturnsReceiverTypeAnnotator(this), super.createTypeAnnotator());
+    }
+
+    /** A TypeAnnotator to add the {@code @}{@link This} annotation. */
+    private class ReturnsReceiverTypeAnnotator extends TypeAnnotator {
+
+        /**
+         * Create a new ReturnsReceiverTypeAnnotator.
+         *
+         * @param typeFactory the {@link AnnotatedTypeFactory} associated with this {@link
+         *     TypeAnnotator}
+         */
+        public ReturnsReceiverTypeAnnotator(AnnotatedTypeFactory typeFactory) {
+            super(typeFactory);
+        }
+
+        @Override
+        public Void visitExecutable(AnnotatedTypeMirror.AnnotatedExecutableType t, Void p) {
+
+            // skip constructors, as we never need to add annotations to them
+            if (!isConstructor(t)) {
+                AnnotatedTypeMirror returnType = t.getReturnType();
+
+                // If any FluentAPIGenerator indicates the method returns this,
+                // add an @This annotation on the return type.
+                if (FluentAPIGenerator.checkForFluentAPIGenerators(t)) {
+                    if (!returnType.isAnnotatedInHierarchy(THIS_ANNOTATION)) {
+                        returnType.addAnnotation(THIS_ANNOTATION);
+                    }
+                }
+
+                // If return type is annotated with @This, add @This annotation
+                // to the receiver type.
+                AnnotationMirror retAnnotation =
+                        returnType.getAnnotationInHierarchy(THIS_ANNOTATION);
+                if (retAnnotation != null
+                        && AnnotationUtils.areSame(retAnnotation, THIS_ANNOTATION)) {
+                    AnnotatedTypeMirror.AnnotatedDeclaredType receiverType = t.getReceiverType();
+                    if (!receiverType.isAnnotatedInHierarchy(THIS_ANNOTATION)) {
+                        receiverType.addAnnotation(THIS_ANNOTATION);
+                    }
+                }
+            }
+            return super.visitExecutable(t, p);
+        }
+    }
+
+    /**
+     * @return {@code true} if the param {@code t} is a {@code Constructor}
+     * @param t a {@link AnnotatedTypeMirror}
+     */
+    private boolean isConstructor(AnnotatedTypeMirror.AnnotatedExecutableType t) {
```

Comment:
There is a single use of this method.
Wouldn't it be much simpler to simply write `if (t.getElement().getKind() == ElementKind.CONSTRUCTOR) { return; }` above? (Addressing my earlier comment about returning early.)